### PR TITLE
do DelegateDel if delegateAdd failed

### LINKED ISSUE #426: flannel /var/lib/cni/flannel/cid  not be removed if cni run failed
/var/lib/cni/flannel/xxx will leak if allocate ip failed.
1.https://github.com/containernetworking/plugins/blob/master/plugins/meta/flannel/flannel.go#L168 allocate ip failed,   
2. https://github.com/containernetworking/plugins/blob/master/plugins/meta/flannel/flannel.go#L164 will leak

maybe add `consumeScratchNetConf`  if  `invoke.DelegateAdd`  failed.

### PR DESCRIPTION
**Pull request by hongli-my**
fix #426

### CODE REVIEW

File: plugins/meta/flannel/flannel_linux.go [hunk 1]
```diff
@@ -65,6 +65,12 @@ func doCmdAdd(args *skel.CmdArgs, n *NetConf, fenv *subnetEnv) error {
 		},
 	}
 
+	defer func() {
+		if err != nil {
+			doCmdDel(args, n)
```
Comment (mars1024) [inline plugins/meta/flannel/flannel_linux.go]: I think this line will never be hit. We need a `defer` in `delegateAdd`.

**Comment by squeed** [comment]
can you add a test for this?

**Comment by bboreham** [comment]
I don't understand why a delegated Del is the right answer here.
Why doesn't the Add side clean up before returning an error?

File: plugins/meta/flannel/flannel_linux.go (hunk 1 above)
Comment (hongli-my) [reply plugins/meta/flannel/flannel_linux.go]: here: https://github.com/containernetworking/plugins/blob/master/plugins/meta/flannel/flannel.go#L170 ?

add here need check `saveScratchNetConf`, `DelegateAdd`, `return result.Print()`  three cases.

Comment (mars1024) [reply plugins/meta/flannel/flannel_linux.go]: no, actually we only need to check `err`

**Comment by dcbw** [comment]
Traditionally the runtime (thing calling CNI) is responsible for cleaning up when ADD fails, eg https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_sandbox.go#L177 because there may be additional things the runtime does that need to be cleaned up outside of networking.

**Pull request closed without merging**